### ISSUE TITLE
Union of generic variants 

### ISSUE THREAD

**Issue by Stouffi** [pelotom/unionize#43]
It would be great if unionize could support union of generic variants

I'm using a generic interface in one of my projects to represent generic fetch status

```typescript
export interface Fetch<T> {
  results: T[]
  finished: boolean
}
```

and I want it to become
```typescript
export interface FetchLoading {
  type: 'loading'
}

export interface FetchSuccess<T> {
  type: 'success'
  results: T[]
  finished: boolean
}

export interface FetchSuccessWithErrors<T> {
  type: 'successWithErrors'
  results: T[]
  finished: boolean
  errors: string[]
}

export interface FetchError {
  type: 'error'
  reason: string
}
```

And I would create an union type which could be consumed by a UI component showing the fetch status according to its type.

How could we achieve that with unionize?

**Comment by OliverJAsh** [comment]
In the absence of support for generics, I'm having to create the tagged union types, constructors, predicates et al. manually. In case it helps anyone else, here is an example: https://gist.github.com/OliverJAsh/9619a49d345c334e5f68473565457167

**Comment by pelotom** [comment]
See [my comment](https://github.com/pelotom/unionize/issues/4#issuecomment-421845259) on #4... it seems like TypeScript needs [generic value types](https://github.com/Microsoft/TypeScript/issues/17574) for this to be practical.

**Comment by twop** [comment]
Hey, I was thinking on that problem a lot and finally shipped a version that Im comfortable with: https://github.com/twop/ts-union
```ts
const Maybe = Union(val => ({
  Nothing: of<void>(),
  Just: of(val)
}));

const {Nothing, Just} = Maybe;

Just(1); // MaybeVal<number>
Just('str'); // MaybeVal<string>
Nothing<boolean>(); // MaybeVal<boolean>
```
Conceptually I have a special token type `Generic` and I replace it with an actual type like `string` at the moment of construction.

```ts
type CreatorFuncG<K, Rec> = K extends Of<infer A>
   ...
    ? A extends [Generic] 
    ? <P>(val: P) => UnionValG<P, Rec> : never
   ...
: never;
```
Might be applicable to `unionize` as well :)

**Comment by OliverJAsh** [comment]
@pelotom Do you think we could work something like what ts-union has into Unionize?

**Comment by pelotom** [comment]
@OliverJAsh I don't have time to work on it myself, but I'm all for it!

**Comment by pelotom** [comment]
@twop any interest in porting your idea into unionize?

**Comment by twop** [comment]
That would be quite a challenge (unionize types a bit more complex), but might be a fun one too :) 

What about API?
```ts
// port from ts-union
const Maybe = unionize(a => ({
  Nothing: ofType<void>(),
  Just: ofType(a)
}),
{value: "val"}); // this is probably required because not every 'a' could be spread
```

I don't have time within the next couple of weeks but happy to try after that

**Comment by pelotom** [comment]
> ```
>   Just: ofType(a) //   'Just: a' might be more ergonomic
> ```

Yeah, I don't see any reason to make it more verbose than it has to be.

>```ts
> {value: "val"}); // this is probably required because not every 'a' could be spread
> ```

Hmm, yeah, I wonder if we want to enforce that they pass a `value` property or just treat it as implied.
